core/src/main/java/org/sonarsource/sonarlint/core/VersionSoonUnsupportedHelper.java

```diff
@@ -0,0 +1,119 @@
+
+@Named
+@Singleton
+public class VersionSoonUnsupportedHelper {
+
+  private static final String UNSUPPORTED_NOTIFICATION_ID = "sonarlint.unsupported.%s.id";
+  private static final String NOTIFICATION_MESSAGE = "The version used by the current connection '%s' will be soon unsupported. " +
+    "Please consider upgrading to the latest %s LTS version to ensure continued support and access to the latest features.";
+  private static final SonarLintLogger LOG = SonarLintLogger.get();
+  private final SonarLintClient client;
+  private final ConfigurationRepository configRepository;
+  private final ConnectionConfigurationRepository connectionRepository;
+  private final ServerApiProvider serverApiProvider;
+
+  public VersionSoonUnsupportedHelper(SonarLintClient client, ConfigurationRepository configRepository,
+    ConnectionConfigurationRepository connectionRepository, ServerApiProvider serverApiProvider) {
+    this.client = client;
+    this.configRepository = configRepository;
+    this.connectionRepository = connectionRepository;
+    this.serverApiProvider = serverApiProvider;
+  }
+
+  @Subscribe
+  public void configurationScopesAdded(ConfigurationScopesAddedEvent event) {
+    var configScopeIds = event.getAddedConfigurationScopeIds();
+    checkIfSoonUnsupported(configScopeIds);
+  }
+
+  @Subscribe
+  public void bindingConfigChanged(BindingConfigChangedEvent event) {
+    var configScopeId = event.getConfigScopeId();
+    var connectionId = event.getNewConfig().getConnectionId();
+    if (connectionId != null) {
+      checkIfSoonUnsupported(configScopeId, connectionId);
+    }
+  }
+
+  private void checkIfSoonUnsupported(Set<String> configScopeIds) {
+    var connectionsPerConfigScopeId = new HashMap<String, String>();
+    configScopeIds.forEach(configScopeId -> {
+      var effectiveBinding = configRepository.getEffectiveBinding(configScopeId);
+      if (effectiveBinding.isPresent()) {
+        var connectionId = effectiveBinding.get().getConnectionId();
+        connectionsPerConfigScopeId.putIfAbsent(connectionId, configScopeId);
+      }
+    });
+    connectionsPerConfigScopeId.forEach((key, value) -> checkIfSoonUnsupported(value, key));
+  }
+
+  private void checkIfSoonUnsupported(String configScopeId, String connectionId) {
+    var connection = connectionRepository.getConnectionById(connectionId);
+    if (connection != null && connection.getKind() == ConnectionKind.SONARQUBE) {
+      serverApiProvider.getServerApi(connectionId).ifPresent(serverApi -> serverApi.system().getStatus()
+          .thenApply(ServerInfo::getVersion)
+          .thenApply(Version::create)
+          .thenApply(VersionUtils::isVersionSupportedDuringGracePeriod)
+          .thenAccept(isSupported -> {
+            if (Boolean.TRUE.equals(isSupported)) {
+              client.showSoonUnsupportedMessage(
+                new ShowSoonUnsupportedMessageParams(
+                  String.format(UNSUPPORTED_NOTIFICATION_ID, connectionId),
```

Comment:
If I understand the ID would be `sonarlint.unsupported.My Connection.id`. There is nothing related to the version itself in the ID. So the following scenario could happen:
* users use a soon-to-be-dropped version, so we send them a notification
* users decide to not show it again
* their SQ version is updated to X.Y
* on a future SL version we mark the X.Y version to be soon dropped
* users won't receive the notification because the ID would be the same and they already dismissed it

I am not sure a SL connection would live long enough for this to happen though

Reply:
Indeed, I knew something was missing 😅